pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java

```diff
@@ -61,6 +62,12 @@ public boolean isInStage() {
         return inStage;
     }
 
+    public void initialize(Map<String, DeclarativeOption> options, boolean explicitAgentInStage) {}
+
+    public boolean useRootAgent(Map<String, DeclarativeOption> options) {
```

Comment:
Not the prettiest signatures; just trying to extract what seems to have been rather ad-hoc existing logic. Suggestions for clearer names and/or parameters welcome, especially something that seems likely to be more generally applicable.

Reply:
The signature of `initialize` seems fine to me. The only thing that comes to mind would be to have `Object context` as a parameter instead of computing `explicitAgentInStage` in the caller in case other implementations want to check some other condition by casting to `Root` or `Stage`, but that's pretty awkward to work with. Besides, given that no other implementations of `DeclarativeAgent` have needed a method like `initialize` yet, it's probably better to keep things simple.

Maybe rename `useRootAgent` to `reuseRootAgent`? I am not very familiar with how everything hooks together to make `newContainerPerStage` work and what exactly the effect is, or if it could make sense for another implementation of `DeclarativeAgent` to ever return true from this method, so I'm not sure about what naming makes the most sense.

Reply:
> casting to `Root` or `Stage`, but that's pretty awkward

Quite awkward, especially as these types are defined in Groovy (for no good reason that I can see).

> I'm not sure about what naming makes the most sense

I really do not know.